airlock/business_logic.py

```diff
@@ -407,10 +410,12 @@ class CommitNotFound(Exception):
     @classmethod
     def from_workspace(cls, workspace: Workspace, commit: str):
         try:
-            repo = workspace.manifest["repo"]
-        except (BusinessLogicLayer.ManifestFileError, KeyError):
+            repo = workspace.manifest.get("repo")
+            if repo is None:
+                repo = list(workspace.manifest["outputs"].values())[0]["repo"]
```

Comment:
I would prefer a different solution here. This API was written before we had manifest.json data, and I think I got it wrong, and it needs a refactor now that we do.

We only ever need a CodeRepo from the context of a specific output file in a workspace.

So I think we should get rid of `from_workspace(workspace: Workspace, commit: str)`, and instead have `from_workspace_output(workspace: Workspace, relpath: UrlPath)`. This looks up the output path in the manifest, and uses the repo/commit metadata found there to construct the CodeRepo object specific for this output. 
This means we don't use the top level `repo` key at all, and we don't need the related job-runner PR.

wdyt?

Reply:
This doesn't entirely work (or at least, it only works for workspace files). This method is used for viewing code for both workspace output files and request files. A manifest only has the latest commit for each workspace file, but a request file could be an older version, based on a previous commit. 

I think we actually want a `from_file(workspace: Workspace, relpath: UrlPath, commit: str = None)` which gets the repo from the manifest for the file and optionally takes a commit (which will have come from a RequestFile), or if it's None, also look it up on the manifest.  

(RequestFileMetadata saves the repo as well as the commit when it's added to a request, but I don't think repo for a workspace should ever change)

Reply:
No, I think it does need it for the views, nvm

Reply:
> This doesn't entirely work (or at least, it only works for workspace files). This method is used for viewing code for both workspace output files and request files. A manifest only has the latest commit for each workspace file, but a request file could be an older version, based on a previous commit.
> 
> I think we actually want a `from_file(workspace: Workspace, relpath: UrlPath, commit: str = None)` which gets the repo from the manifest for the file and optionally takes a commit (which will have come from a RequestFile), or if it's None, also look it up on the manifest.

How about two constructors?

`from_workspace(workspace, relpath)`

and

`from_request(release_request, relpath)`

> (RequestFileMetadata saves the repo as well as the commit when it's added to a request, but I don't think repo for a workspace should ever change)

It could in theory, if the repo is renamed by an admin during the lifetime of the project (I think this has happened). Which is one of the reasons to take it from the output metadata.  But its a very edge case.

Also, multiple workspaces can share the same repo, but with different branches.  Here, we only ever care about a specific commit, so we don't need branch at all.

Reply:
Right, I've spent a while trying to figure this out, and i think it may have to stay as `from_workspace`.  It's perfectly straightforward to get the repo from an output or request file by looking up a relpath in the manifest. The problem is that it's only used in the code views, which have a path, but it's not the output/request path that the code comes from, it's the path to the file within the code repo. 

I can't put both the path of the code repo file and the path of the source file in the URL; I tried parsing the return_url, from which you can get the source relpath; but then you also need to pass it on to the contents view, and deal with the htmx requests that don't have the GET params; or you to have have a way to retrieve the repo without it, which ends up back at getting the repo from the first output file.

I can make it always retrieve the repo from the first output file though; they should always be the same for all outputs, and we shouldn't ever be trying to get a repo without an output file - so the job-runner PR still isn't required.

Reply:
> How about two constructors?

No, still doesn't work with the code views. Retrieving it from the first output file for both workspace and request files (as it does now) handles the issue of a repo changing name, and both pass the commit when they build the code url. I think the current method is still the best way of doing this.

Reply:
> Right, I've spent a while trying to figure this out, and i think it may have to stay as `from_workspace`. It's perfectly straightforward to get the repo from an output or request file by looking up a relpath in the manifest. The problem is that it's only used in the code views, which have a path, but it's not the output/request path that the code comes from, it's the path to the file within the code repo.

> I can't put both the path of the code repo file and the path of the source file in the URL; I tried parsing the return_url, from which you can get the source relpath; but then you also need to pass it on to the contents view, and deal with the htmx requests that don't have the GET params; or you to have have a way to retrieve the repo without it, which ends up back at getting the repo from the first output file.

Ah, right, the current url structure bakes in the assumption of a workspace -> repo mapping.  I think maybe that is the issue - we should probably have the repo name in the url, rather than workspace? So we can just resolve it once for the output in question, and just re-use it for browsing after that? Something for later on.

> I can make it always retrieve the repo from the first output file though; they should always be the same for all outputs, and we shouldn't ever be trying to get a repo without an output file - so the job-runner PR still isn't required.

Right, that seems like a reasonable compromise. Sorry for sending you on a wild goose chase here :(

Reply:
So yes, I also thought about having repo name in the URL instead of workspace, but then it's not straightforward to get the repo URL without the workspace (which has the manifest, with the full url). Presumably it should always be https://github.com/opensafely/<reponame> (except in tests) - but i think using the workspace is still probably more reliable. For now at least

Reply:
Updated now to always use the first output; plus a change to the request_return functional test because it had started failing locally for me with a database locked error on this branch. Hoping that's also the fix for the intermittent CI errors too.

Reply:
Logging out is the fix to the db issue? That's interesting.  Something about sessions maybe?